sdks/python/src/redactive/multi_user_client.py

```diff
@@ -0,0 +1,138 @@
+from redactive.grpc.v1 import Chunk, RelevantChunk
+from redactive.search_client import SearchClient
+
+
+@dataclass
+class UserData:
+    refresh_token: str | None = None
+    id_token: str | None = None
+    id_token_expiry: datetime | None = None
+    connections: list[str] = field(default_factory=list)
+    sign_in_state: str | None = None
+
+
+@dataclass
+class MultiUserClientConfig:
+    auth_base_url: str
+    """The base auth URL to use for Redactive."""
+    grpc_host: str
+    """The host to use for the gRPC server for Search services."""
+    grpc_port: int
+    """The port to use for the gRPC server for Search services."""
+
+
+class InvalidRedactiveSessionError(Exception):
+    def __init__(self, user_id: str) -> None:
+        super().__init__(f"No valid Redactive session for user '{user_id}'")
+
+
+class MultiUserClient:
+    def __init__(
+        self,
+        api_key: str,
+        callback_uri: str,
+        read_user_data: Callable[[str], Awaitable[UserData]],
+        write_user_data: Callable[[str, UserData | None], Awaitable[None]],
+        config: MultiUserClientConfig,
+    ) -> None:
+        """Redactive client handling multiple users authentication and access to the Redactive Search service.
+
+        :param api_key: Redactive API key.
+        :type api_key: str
+        :param callback_uri: The URI to redirect to after initiating the connection. Defaults to an empty string.
+        :type callback_uri: str
+        :param read_user_data: Function to read user data from storage.
+        :type read_user_data: Callable[[str], Awaitable[UserData]]
+        :param write_user_data: Function to write user data to storage.
+        :type write_user_data: Callable[[str, UserData | None], Awaitable[None]]
+        :param config: configuration of the auth client and the search client
+        :type config: MultiUserClientConfig
+        """
+
+        self.auth_client = AuthClient(api_key, base_url=config.auth_base_url)
+        self.search_client = SearchClient(config.grpc_host, config.grpc_port)
+        self.callback_uri = callback_uri
+        self.read_user_data = read_user_data
+        self.write_user_data = write_user_data
+
+    async def get_begin_connection_url(self, user_id: str, provider: str) -> str:
+        state = str(uuid.uuid4())
+        response = await self.auth_client.begin_connection(provider, self.callback_uri, code_param_alias=state)
+        user_data: UserData = await self.read_user_data(user_id)
+        user_data.sign_in_state = state
+        await self.write_user_data(user_id, user_data)
+        return response.url
+
+    async def _refresh_user_data(
+        self, user_id: str, refresh_token: str | None = None, sign_in_code: str | None = None
+    ) -> UserData:
+        tokens = await self.auth_client.exchange_tokens(sign_in_code, refresh_token)
+        connections: ListConnectionsResponse = await self.auth_client.list_connections(tokens.idToken)
+        user_data: UserData = UserData(
```

Comment:
```suggestion
        user_data = UserData(
```
I think a type-hint here is unnecessary if the line is creating that type